server/fleet/microsoft_mdm.go

```diff
@@ -820,77 +823,373 @@ func (e MDMWindowsEnrolledDevice) AuthzType() string {
 }
 
-type SyncMLCommand struct {
-	XMLName xml.Name
-	CmdID   int                 `xml:",omitempty"`
-	MsgRef  string              `xml:",omitempty"`
-	CmdRef  string              `xml:",omitempty"`
-	Cmd     string              `xml:",omitempty"`
-	Target  string              `xml:"Target>LocURI"`
-	Source  string              `xml:"Source>LocURI"`
-	Data    string              `xml:",omitempty"`
-	Item    []SyncMLCommandItem `xml:",any"`
+// Protocol Command
+type SyncMLCmd struct {
+	XMLName xml.Name   `xml:",omitempty"`
+	CmdID   string     `xml:"CmdID"`
+	MsgRef  *string    `xml:"MsgRef,omitempty"`
+	CmdRef  *string    `xml:"CmdRef,omitempty"`
+	Cmd     *string    `xml:"Cmd,omitempty"`
+	Data    *string    `xml:"Data,omitempty"`
+	Items   *[]CmdItem `xml:"Item,omitempty"`
 }
 
-type SyncMLBody struct {
-	Item []SyncMLCommand `xml:",any"`
+type CmdItem struct {
+	Source *string `xml:"Source>LocURI,omitempty"`
+	Target *string `xml:"Target>LocURI,omitempty"`
+	Meta   *Meta   `xml:"Meta,omitempty"`
+	Data   *string `xml:"Data"`
 }
 
-// IsValidSyncMLMsg checks for required fields in the SyncML message
-func (req *SyncMLMessage) IsValidSyncMLMsg() error {
-	if req == nil {
-		return errors.New("invalid SyncML message: nil")
+type Meta struct {
+	Type   *MetaAttr `xml:"Type,omitempty"`
+	Format *MetaAttr `xml:"Format,omitempty"`
+}
+
+type MetaAttr struct {
+	XMLNS   string  `xml:"xmlns,attr"`
+	Content *string `xml:",chardata"`
+}
+
+type LocURI struct {
+	LocURI *string `xml:",omitempty"`
+}
+
+func (msg *SyncML) IsValidHeader() error {
+	if strings.TrimSpace(msg.Xmlns) == "" {
+		return errors.New("msg namespace")
+	}
+
+	// SyncML DTD version check
+	if msg.SyncHdr.VerDTD != mdm.SyncMLSupportedVersion {
+		return errors.New("unsupported DTD version")
 	}
 
-	if len(req.Header.Version) == 0 {
-		return errors.New("invalid SyncML message: Version")
+	// SyncML Proto version check
+	if msg.SyncHdr.VerProto != mdm.SyncMLVerProto {
+		return errors.New("unsupported proto version")
 	}
 
-	if len(req.Header.Target) == 0 {
-		return errors.New("invalid SyncML message: Target")
+	// SyncML SessionID check
+	if utf8.RuneCountInString(msg.SyncHdr.SessionID) < 1 {
+		return errors.New("sessionID")
 	}
 
-	if req.Header.SessionID == 0 {
-		return errors.New("invalid SyncML message: SessionID")
+	// SyncML MsgID check
+	if utf8.RuneCountInString(msg.SyncHdr.MsgID) < 1 {
+		return errors.New("MsgID")
 	}
 
-	if req.Header.MsgID == 0 {
-		return errors.New("invalid SyncML message: SessionID")
+	// Target LocURI check
+	if strings.TrimSpace(*msg.SyncHdr.Target.LocURI) == "" {
+		return errors.New("Target.LocURI")
 	}
 
-	if len(req.Body.Item) == 0 {
-		return errors.New("invalid SyncML message: Item")
+	// Device ID check
+	if strings.TrimSpace(*msg.SyncHdr.Source.LocURI) == "" {
+		return errors.New("Source.LocURI")
 	}
 
 	return nil
 }
+
+func (msg *SyncML) IsValidBody() error {
+	nonNilCount := 0
+
+	if msg.SyncBody.Add != nil {
```

Comment:
Assuming you change the `ProtoCmds` type to be a slice (and not a pointer to a slice), I think the check should be `if len(msg.SyncBody.Add) > 0 {` (same for each command below) because the slice could be nil but it could also be empty, meaning that no command was provided.

Reply:
I updated this to reflect ProtoCmds being a slice.